### PR TITLE
B-21488: UB Allowance Warning

### PR DESCRIPTION
**Pull request by TevinAdams**
## Summary

In this PR implementation is being added to evaluate excess weight for a UB shipment. Excess weight can be evaluated on shipments at 2 key points. When a shipment is approved and when a shipment is updated. On approval if the prime estimated weight exist then it will be evaluated. On update if the updated shipment has a prime estimated weight included.

For UB shipments, based on the AC of the story, when the primeEstimatedWeight or the primeActualWeight is within 10% or greater of the UBAllowance then excessWeight will be triggered until it has been acknowledged. Same flow of all other shipments.

Is there anything you would like reviewers to give additional scrutiny?

[this article](tbd) explains more about the approach used.

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Ensure FEATURE_FLAG_UNACCOMPANIED_BAGGAGE is set to true
2. Determine a duty location you will use when creating a move. In the DB set that duty location address is_Oconus value to true
3.Create a move using that address discussed above as one of the duty locations
4. Create a ub shipment and add dependents
5. Complete the move creation process in milmove
6. As SC complete the counsel process for the move
7. As TOO approve the move. (Here excess weight will not be evaluated because primeEstimatedWeight is null at this time)
8. Look in the devs tools under entitlement object and take note of the ubAllowance. (Lets say it is 2000)
9. As Prime in the prime sim update the shipment adding a estimated weight that will trigger the excess weight(Lets say 1850). For ubAllowance of 2000 the threshold value is 1800. 2000 * .9 = 1800.
10. After updating the shipment view the move as TOO in the move queue. The status should say Approvals Requested.
image
11. Next as TOO view the move and under the Move Task Order tab and you should see the excess weight warning.
image
12. Acknowledge the warning by clicking the "Review billable weight" link. You should notice the warning disappear
image
image
13. Next view the move in the move queue and you should see the status go back to approved
image

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

### CODE REVIEW

**Comment by cameroncaci** [comment]
@TevinAdams I think your PR description and such forgot to save when you opened it up, marking back to draft

**Comment by cameroncaci** [comment]
Please reopen the PR as ready once the description and test steps are refreshed

**Comment by TevinAdams** [comment]
> @TevinAdams I think your PR description and such forgot to save when you opened it up, marking back to draft

@cameroncaci I was still in the process of building out the PR. Its done now.

File: pkg/services/move/move_weights.go [hunk 1]
```diff
@@ -132,8 +132,26 @@ func (w moveWeights) CheckExcessWeight(appCtx appcontext.AppContext, moveID uuid
...
+	if updatedShipment.ShipmentType == models.MTOShipmentTypeUnaccompaniedBaggage && move.Orders.Entitlement.UBAllowance != nil {
+		ubExcessPrimeEstimatedWeight := false
+		if updatedShipment.PrimeEstimatedWeight != nil {
+			ubExcessPrimeEstimatedWeight = float64(*move.Orders.Entitlement.UBAllowance)*RiskOfExcessThreshold <= float64(*updatedShipment.PrimeEstimatedWeight)
+		}
+
+		ubExcessPrimeActualWeight := false
+		if updatedShipment.PrimeActualWeight != nil {
+			ubExcessPrimeActualWeight = float64(*move.Orders.Entitlement.UBAllowance)*RiskOfExcessThreshold <= float64(*updatedShipment.PrimeActualWeight)
+		}
```
Comment (cameroncaci) [inline pkg/services/move/move_weights.go:146]: Missing code coverage here
Screenshot 2024-11-21 at 1 37 57 PM

**Review by cameroncaci — CHANGES_REQUESTED** [review]
Identified frontend issue & a missed code coverage spot.

Verified using actual and estimated via postman as well as prime sim too

https://github.com/user-attachments/assets/32445203-326e-46b7-bd98-e5c604a67e96

**Comment by danieljordan-caci** [comment]
Are we sure this doesn't need to be its own allowance check? Did we verify that they are cool with it being shown in the move history as the same action of the TOTAL weight allowance being acknowledged? Without different columns in the db then we won't know if the UB allowance or the total weight allowance was in excess.

**Comment by danieljordan-caci** [comment]
For example, if I have a UB shipment & an HHG shipment.

I updated the est allowance to trigger the UB excess weight and get the banner, but there's no way to know if it's the total weight allowance or the UB allowance

Screenshot 2024-11-21 at 1 10 44 PM

**Comment by traskowskycaci** [comment]
> For example, if I have a UB shipment & an HHG shipment.
> 
> I updated the est allowance to trigger the UB excess weight and get the banner, but there's no way to know if it's the total weight allowance or the UB allowance
> 
> Screenshot 2024-11-21 at 1 10 44 PM

Had the same question - and do we need to show the UB allowance itself on this MTO page as part the excess weight? The office users can see it on the Move details page but they probably don't want to toggle back and forth

**Comment by danieljordan-caci** [comment]
> > For example, if I have a UB shipment & an HHG shipment.
> > I updated the est allowance to trigger the UB excess weight and get the banner, but there's no way to know if it's the total weight allowance or the UB allowance
> > Screenshot 2024-11-21 at 1 10 44 PM
> 
> Had the same question - and do we need to show the UB allowance itself on this MTO page as part the excess weight? The office users can see it on the Move details page but they probably don't want to toggle back and forth

It's pretty confusing to the user. I think we need to let them know "hey they're UB shipment allowance is in excess" so we can 1. be able to audit that and 2. the office user isn't confused that the weight allowance is actually fine, it's just the UB allowance that's at risk

**Comment by traskowskycaci** [comment]
> > > For example, if I have a UB shipment & an HHG shipment.
> > > I updated the est allowance to trigger the UB excess weight and get the banner, but there's no way to know if it's the total weight allowance or the UB allowance
> > > Screenshot 2024-11-21 at 1 10 44 PM
> > 
> > 
> > Had the same question - and do we need to show the UB allowance itself on this MTO page as part the excess weight? The office users can see it on the Move details page but they probably don't want to toggle back and forth
> 
> It's pretty confusing to the user. I think we need to let them know "hey they're UB shipment allowance is in excess" so we can 1. be able to audit that and 2. the office user isn't confused that the weight allowance is actually fine, it's just the UB allowance that's at risk

I also brought this up to Wendy earlier - so far, our work has been to calculate UB allowance in isolation, but it is _part_ of the total move allowance. As far as I know, we aren't doing any math to subtract the UB allowance from the TOTAL move weight allowance - something we need to think about and may work in the future. Making sure we distinctly show UB excess weight vs move excess weight will help clarify some of that math we might need to do.

**Comment by cameroncaci** [comment]
@traskowskycaci @danieljordan-caci 

I voiced these same exact concerns and Dan wrote it off as okay (Verbally in the past) by just sharing the same exact flow. It would be better UX to separate them though. Since Dan is out we may have to have Terri write it off as okay in Agility

**Comment by TevinAdams** [comment]
> > @traskowskycaci @danieljordan-caci
> > I voiced these same exact concerns and Dan wrote it off as okay (Verbally in the past) by just sharing the same exact flow. It would be better UX to separate them though. Since Dan is out we may have to have Terri write it off as okay in Agility
> 
> I agree. I do think we should see if she's available to meet for a second after standup and demo what we are talking about and how the ambiguity can be confusing to the end user. I can pull branch and have it to demo if she's available.

@danieljordan-caci, @cameroncaci , @traskowskycaci   I am okay with a seperate message for UB I was just following the the approach of all the other shipments that doesn't have a specific message. Seemed like the AC wanted to use what was already there.

**Comment by cameroncaci** [comment]
@TevinAdams per that meeting we had, marking this PR back to draft until changes implemented

**Pull request closed without merging**